plugins/bil_pipe/bil_pipe_main.ml

```diff
@@ -0,0 +1,62 @@
+open Core_kernel
+open Bap.Std
+open Bil_const_propagation
+open Bil_renum_virtual
+open Bil_memo
+include Self ()
+
+let parse = function
+  | "enable" -> `Ok true
+  | "disable" -> `Ok false
+  | s ->
+    `Error (sprintf "unknown value %s, possible are: disable | enable" s)
+
+let print fmt s =
+  Format.fprintf fmt "%s" @@
+  match s with
+  | true  -> "enable"
+  | false -> "disable"
+
+let switcher ?(default=true) =
+  Config.(param (converter parse print default) ~default)
+
+let simpl = Bil.fold_consts
+let norml = Stmt.normalize ~keep_ites:false ~normalize_exp:false
+let propg = propagate_consts
+
+let process pipe =
+  let pipe = List.filter_map pipe
+      ~f:(fun (test, f) -> Option.some_if test f) in
+  fun bil ->
+    let rec run bil = function
+      | [] -> bil
+      | f :: fs ->
+        run (f bil) fs in
+    run bil pipe
+
+let run if_norml if_simpl if_propg =
+  register_bass "internal" @@
+  process [
+    if_simpl,simpl;
+    if_norml,norml;
+    if_propg,propg;
+  ]
+
+let () =
+  let () = Config.manpage [
+      `S "DESCRIPTION";
+      `P "Applies analysises to a instruction bil code" ;
+      `Pre "
+The following bil analysises are in default pipeline:
+Constant Folding and Simplification
+Bil Normalization
+Constant Propagation
+Memoization
```

Comment:
Though I don't see any memoizations here, you can actually help yourself with a hashtable that holds simplified BIL. You can compute a digest of a BIL code it should be faster than applying all your optimizations.

The main consideration here, is that we will leak a lot of memory in such memoization. Thus you shall implement some sort of cache eviction mechanism, e.g., the intuitive suggestion would be:
1) do not store trivial instructions that are not simplified
2) store a fixed number of instructions, say a hundred
3) once the limit is reached evict a random instruction